model/generate.go

```diff
@@ -1,25 +1,296 @@
+			}
+		} else {
+			// If the buildvariant does not exist, create it.
+			intermediateProject.BuildVariants = append(intermediateProject.BuildVariants, bv)
+		}
+	}
+	byteConfig, err := yaml.Marshal(intermediateProject)
+	if err != nil {
+		return "", errors.Wrap(err, "error marshalling new project config")
+	}
+	return string(byteConfig), nil
+}
+
+// projectMaps is a struct of maps of project fields, which allows efficient comparisons of generated projects to projects.
+type projectMaps struct {
+	buildVariants map[string]struct{}
+	tasks         map[string]*ProjectTask
+	functions     map[string]*YAMLCommandSet
+}
+
+// validateMaxTasksAndVariants validates that the GeneratedProject contains fewer than 100 variants and 1000 tasks.
+func (g *GeneratedProject) validateMaxTasksAndVariants(catcher grip.Catcher) {
+	if len(g.BuildVariants) > maxGeneratedBuildVariants {
+		catcher.Add(errors.Errorf("it is illegal to generate more than %d buildvariants", maxGeneratedBuildVariants))
+	}
+	if len(g.Tasks) > maxGeneratedTasks {
+		catcher.Add(errors.Errorf("it is illegal to generate more than %d tasks", maxGeneratedTasks))
+	}
+}
+
+// validateNoRedefine validates that buildvariants, tasks, or functions, are not redefined
+// except to add a task to a buildvariant.
+func (g *GeneratedProject) validateNoRedefine(cachedProject projectMaps, catcher grip.Catcher) {
+	for _, bv := range g.BuildVariants {
+		if _, ok := cachedProject.buildVariants[bv.Name]; ok {
+			if bv.DisplayName != "" || len(bv.Expansions) > 0 || len(bv.Modules) > 0 ||
+				bv.Disabled == true || len(bv.Tags) > 0 || bv.Push == true ||
+				bv.BatchTime != nil || bv.Stepback != nil || len(bv.RunOn) > 0 ||
+				len(bv.DisplayTasks) > 0 {
+				catcher.Add(errors.Errorf("cannot redefine buildvariants in 'generate.tasks' (%s), except to add tasks", bv.Name))
+			}
+		}
+	}
+	for _, t := range g.Tasks {
+		if _, ok := cachedProject.tasks[t.Name]; ok {
+			catcher.Add(errors.Errorf("cannot redefine tasks in 'generate.tasks' (%s)", t.Name))
+		}
+	}
+	for f := range g.Functions {
+		if _, ok := cachedProject.functions[f]; ok {
+			catcher.Add(errors.Errorf("cannot redefine functions in 'generate.tasks' (%s)", f))
+		}
+	}
+}
+
+// validateNoRecursiveGenerateTasks validates that no 'generate.tasks' calls another 'generate.tasks'.
+func (g *GeneratedProject) validateNoRecursiveGenerateTasks(cachedProject projectMaps, catcher grip.Catcher) {
+	for _, t := range g.Tasks {
+		for _, cmd := range t.Commands {
+			if cmd.Command == "generate.tasks" {
+				catcher.Add(errors.New("cannot define 'generate.tasks' from a 'generate.tasks' block"))
+			}
+		}
+	}
+	for _, f := range g.Functions {
+		for _, cmd := range f.List() {
+			if cmd.Command == "generate.tasks" {
+				catcher.Add(errors.New("cannot define 'generate.tasks' from a 'generate.tasks' block"))
+			}
+		}
+	}
+	for _, bv := range g.BuildVariants {
+		for _, t := range bv.Tasks {
+			if projectTask, ok := cachedProject.tasks[t.Name]; ok {
```

Comment:
Can you refactor this block into its own function to reduce the number of embedded stack levels here?

Reply:
done